direct_access.cpp

```diff
@@ -0,0 +1,133 @@
+
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <sstream>
+
+#include "moab/Core.hpp"
+
+#include "timer.h"
+
+int main(int argc, char** argv){
+
+  std::shared_ptr<moab::Interface> MBI = std::make_shared<moab::Core>();
+
+
+  // load file
+  std::cout << "Loading MOAB model...";
+  moab::ErrorCode rval = MBI->load_file("dagmc.h5m");
+  MB_CHK_SET_ERR(rval, "Failed to load the file");
+  std::cout << "done." << std::endl;
+
+  // get all the triangles in the model
+  moab::Range all_tris;
+  rval = MBI->get_entities_by_type(0, moab::MBTRI, all_tris);
+  MB_CHK_SET_ERR(rval, "Failed to get all triangles");
+
+  std::cout << "Model contains " << all_tris.size() << " triangles." << std::endl;
+
+  // get direct access pointers
+  moab::EntityHandle* conn_ptr;
+  moab::EntityHandle first_element = all_tris.front();
+  int stride, n_tris;
+  rval = MBI->connect_iterate(all_tris.begin(), all_tris.end(), conn_ptr, stride, n_tris);
+  MB_CHK_SET_ERR(rval, "Failed to get connectivity pointer");
+
+  double* x_vals;
+  double* y_vals;
+  double* z_vals;
+
+  moab::Range all_verts;
+  rval = MBI->get_entities_by_dimension(0, 0, all_verts);
+  MB_CHK_SET_ERR(rval, "Failed to get all vertices in the model");
+
+  int n_vertices;
+  rval = MBI->coords_iterate(all_verts.begin(), all_verts.end(), x_vals, y_vals, z_vals, n_vertices);
+  MB_CHK_SET_ERR(rval, "Failed to get coordinate pointers");
+
+  size_t n_samples = 10000000;
+
+  // write binary file if indicated by "-w" flag
+  bool write_vals = argc > 1 && std::string(argv[1]) == "-w";
+
+  Timer t;
+  std::cout << "Running " << n_samples << " samples." << std::endl;
+
+  // vector holding a random set of indices to use as samples
+  std::vector<size_t> random_indices(n_samples);
+  srand(100); // ensure random number sequence is the same
+  for(int i = 0; i < n_samples; i++) {
+    random_indices.at(i) = rand() % all_tris.size();
+  }
+
+  std::vector<std::string> results(n_samples);
+
+  t.start();
+  #pragma omp parallel for shared(MBI, write_vals, random_indices, results)
+  for(size_t i = 0; i < n_samples; i++) {
+
+    // get random index for this sample
+    size_t idx = random_indices[i];
+
+    // query the triangle connectivity
+    moab::EntityHandle eh = all_tris[idx];
+
+    moab::Range conn;
+    size_t offset = stride * (eh - all_tris.front());
+
+    moab::EntityHandle v0 = conn_ptr[offset];
+    moab::EntityHandle v1 = conn_ptr[offset + 1];
+    moab::EntityHandle v2 = conn_ptr[offset + 2];
```

Comment:
The code below assumes that first vertex handle is 1, which is in some ways ok, but not universally applicable. If you're sure your handles are in a single run, it's safer to use:

for (unsigned i = 0; i < 3; ++i) vidx[i] = connptr[3*idx+i] - firstVert;

Then the coords for vertex i are x_vals[vidx[i]], y_vals[vidx[i]] and z_vals[vidx[i]].